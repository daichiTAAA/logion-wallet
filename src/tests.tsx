import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };
