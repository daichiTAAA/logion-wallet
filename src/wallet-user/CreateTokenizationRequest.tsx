import React from 'react';
import { useForm, Controller } from 'react-hook-form';
import Form from "react-bootstrap/Form";
import ButtonGroup from "react-bootstrap/ButtonGroup";
import Button from "react-bootstrap/Button";
import moment from 'moment';

import {CreateTokenRequest} from "./Model";
import {useUserContext} from "./UserContext";
import { sign } from '../logion-chain';

export interface Props {
    onSubmit: () => void;
    onCancel: () => void;
}

export default function CreateTokenizationRequest(props: Props) {

    interface FormValues {
        requestedTokenName: string,
        bars: number,
    }

    const { control, handleSubmit, formState: { errors } } = useForm<FormValues>();
    const { legalOfficerAddress, userAddress, createTokenRequest } = useUserContext();

    const submit = async (formValues: FormValues) => {
        const attributes = [
            `${legalOfficerAddress}`,
            `${formValues.requestedTokenName}`,
            `${formValues.bars}`
        ];
        const signedOn = moment();
        const signature = await sign({
            signerId: userAddress,
            resource: 'token-request',
            operation: 'create',
            signedOn,
            attributes,
        });
        const request: CreateTokenRequest = {
            legalOfficerAddress: legalOfficerAddress,
            requesterAddress: userAddress,
            bars: Number(formValues.bars),
            requestedTokenName: formValues.requestedTokenName,
            signature,
            signedOn,
        }
        await createTokenRequest!(request);
        props.onSubmit();
    }

    return (
        <Form onSubmit={handleSubmit(submit)}>
            <Form.Group controlId="requestedTokenName">
                <Form.Label>Token Name</Form.Label>
                <Controller
                    name="requestedTokenName"
                    control={control}
                    defaultValue=""
                    rules={{
                            required: 'The token name is required',
                            minLength: {
                                value: 3,
                                message: 'The token name must contain at least 3 characters'
                            },
                            maxLength: {
                                value: 40,
                                message: 'The token name must contain at most 40 characters'
                            }
                    }}
                    render={({ field }) => <Form.Control isInvalid={!!errors.requestedTokenName?.message} type="text" placeholder="e.g. XYZ" data-testid="tokenName" {...field} />}
                  />
                  <Form.Control.Feedback type="invalid" data-testid="tokenNameMessage">{errors.requestedTokenName?.message}</Form.Control.Feedback>
            </Form.Group>

            <Form.Group controlId="bars">
                <Form.Label>Number of Gold Bars</Form.Label>
                <Controller
                    name="bars"
                    control={control}
                    defaultValue=""
                    rules={{
                        required: 'The # of bars is required',
                        min: {
                            value: 1,
                            message: 'The # of bars must be greater or equal to 1'
                        },
                        max: {
                            value: 100,
                            message: 'The # of bars must not be greater than 100'
                        }
                    }}
                    render={({ field }) => <Form.Control isInvalid={!!errors.bars?.message} type="number" placeholder="e.g. 3" data-testid="bars" {...field} />}
                  />
                  <Form.Control.Feedback type="invalid" data-testid="barsMessage">{errors.bars?.message}</Form.Control.Feedback>
            </Form.Group>

            <ButtonGroup>
                <Button type="submit" variant="primary" data-testid="btnSubmit">Submit</Button>
                <Button variant="secondary" data-testid="btnCancel" onClick={props.onCancel}>Cancel</Button>
            </ButtonGroup>
        </Form>
    )
}
