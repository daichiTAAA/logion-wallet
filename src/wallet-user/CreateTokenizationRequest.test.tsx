import React from 'react';
import CreateTokenizationRequest from "./CreateTokenizationRequest";
import {shallowRender} from "../tests";
import {act, fireEvent, render} from '@testing-library/react';

test("renders", () => {
    const tree = shallowRender(<CreateTokenizationRequest onCancel={() => null} onSubmit={() => null}/>)
    expect(tree).toMatchSnapshot();
});

test("submit an empty form", () => {
    const cancelCallback = jest.fn();
    const submitCallback = jest.fn();
    const tree = render(<CreateTokenizationRequest onCancel={cancelCallback} onSubmit={submitCallback}/>)
    const button = tree.getByTestId("btnSubmit");
    act(() => { fireEvent.click(button) });
    expect(cancelCallback).not.toBeCalled();
    expect(submitCallback).not.toBeCalled();
});

/*
test("submit a valid form", () => {
    const cancelCallback = jest.fn();
    const submitCallback = jest.fn();
    const tree = render(<CreateTokenizationRequest onCancel={cancelCallback} onSubmit={submitCallback}/>)

    const tokeName = tree.getByTestId("tokenName");
    const bars = tree.getByTestId("bars");
    const button = tree.getByTestId("btnSubmit");
    act(() => {
        fireEvent.input(tokeName, {target: {
            value: 'DOT7'
            }})
        fireEvent.input(bars, {target: {
            value: '7'
            }})
        fireEvent.click(button);
    });
    expect(cancelCallback).not.toBeCalled();
    expect(submitCallback).toBeCalled();
    // TODO Fix this test - see https://www.react-hook-form.com/advanced-usage#TestingForm
});

 */

test("cancel", () => {
    const cancelCallback = jest.fn();
    const submitCallback = jest.fn();
    const tree = render(<CreateTokenizationRequest onCancel={cancelCallback} onSubmit={submitCallback}/>)
    const button = tree.getByTestId("btnCancel");
    act(() => { fireEvent.click(button) });
    expect(cancelCallback).toBeCalled();
    expect(submitCallback).not.toBeCalled();
});

