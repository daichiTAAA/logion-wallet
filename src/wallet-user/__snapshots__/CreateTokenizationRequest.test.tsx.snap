// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    data-testid="tokenName"
    name="requestedTokenName"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Token name"
    type="text"
  />
  <p />
  <input
    data-testid="bars"
    name="bars"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Number of Gold Bars"
    type="number"
  />
  <p />
  <Button
    active={false}
    data-testid="btnSubmit"
    disabled={false}
    type="submit"
    variant="primary"
  >
    Submit
  </Button>
  <Button
    active={false}
    data-testid="btnCancel"
    disabled={false}
    onClick={[Function]}
    variant="primary"
  >
    Cancel
  </Button>
</form>
`;
